src/sage/rings/lazy_species.py

```diff
@@ -0,0 +1,486 @@
+    r"""
+    Return all compositions of the vector `n` of weight `d`.
+
+    INPUT:
+
+    - ``n_vec``, a `k`-tuple of non-negative integers.
+
+    - ``d``, a non-negative integer.
+
+    - ``weights_vec``, `k`-tuple of weakly increasing lists of
+      positive integers.
+
+    EXAMPLES::
+
+        sage: from sage.rings.lazy_species import weighted_vector_compositions
+        sage: list(weighted_vector_compositions([1,1], 2, [[1,1,2,3], [1,2,3]]))
+        [([0, 1], [1]), ([1], [1])]
+
+        sage: list(weighted_vector_compositions([3,1], 4, [[1,1,2,5], [1,1,2,5]]))
+        [([0, 3], [0, 1]),
+         ([0, 3], [1]),
+         ([1, 2], [0, 1]),
+         ([1, 2], [1]),
+         ([2, 1], [0, 1]),
+         ([2, 1], [1]),
+         ([3], [0, 1]),
+         ([3], [1])]
+
+    """
+    k = len(n_vec)
+    for d_vec in IntegerVectors(d, length=k):
+        yield from itertools.product(*map(weighted_compositions, n_vec, d_vec, weights_vec))
+
+
+######################################################################
+
+class LazySpeciesElement(LazyCompletionGradedAlgebraElement):
+    r"""
+    Compute the molecular expansion of `E(-X)`::
+
+        sage: from sage.rings.lazy_species import LazySpecies
+        sage: L = LazySpecies(ZZ, "X")
+        sage: E = L(lambda n: SymmetricGroup(n))
+        sage: 1 / E
+        1 + (-X) + (-E_2+X^2) + (-E_3+2*X*E_2-X^3)
+          + (-E_4+2*X*E_3+E_2^2-3*X^2*E_2+X^4)
+          + (-E_5+2*X*E_4+2*E_2*E_3-3*X^2*E_3-3*X*E_2^2+4*X^3*E_2-X^5)
+          + (-E_6+2*X*E_5+2*E_2*E_4-3*X^2*E_4+E_3^2-6*X*E_2*E_3+4*X^3*E_3-E_2^3+6*X^2*E_2^2-5*X^4*E_2+X^6)
+          + O^7
+
+    Compare with the explicit formula::
+
+        sage: def coefficient(m):
+        ....:     return sum((-1)^len(la) * multinomial((n := la.to_exp())) * prod(E[i]^ni for i, ni in enumerate(n, 1)) for la in Partitions(m))
+
+        sage: all(coefficient(m) == (1/E)[m] for m in range(10))
+        True
+    """
+    def generating_series(self):
+        pass
+
+    def isotype_generating_series(self):
+        pass
+
+    def cycle_index_series(self):
+        pass
+
+    def _add_(self, other):
+        r"""
+        Return the sum of ``self`` and ``other``.
+
+        In particular, the method to obtain the structures is
+        adapted.
+
+        EXAMPLES::
+
+            sage: from sage.rings.lazy_species import LazySpecies
+            sage: L = LazySpecies(ZZ, "X")
+            sage: E = L(lambda n: SymmetricGroup(n))
+            sage: F = L(lambda n: SymmetricGroup(n))
+            sage: list(E.structures([1,2,3]))
+            [((1, 2, 3), E_3)]
+            sage: list((E+F).structures([1,2,3]))
+            [((1, 2, 3), E_3), ((1, 2, 3), E_3)]
+
+        """
+        def add_structures(labels):
+            yield from self.structures(labels)
+            yield from other.structures(labels)
+
+        result = super()._add_(other)
+        result.structures = add_structures
+        return result
```

Comment:
@tscrim, I'm not sure whether this is the best way to do it.  I tried to create a separate class `SumSpeciesElement` inheriting from `LazySpeciesElement` that only overrides the method `structures` (and will override `generating_series`, etc.), but got stuck.

I think I would like to turn the result of `super()._add_(other)` into a `SumSpeciesElement` instance.

Reply:
One other thing: I am not sure, but perhaps it would actually be nice to take some decisions of `LazyModuleElement._add_` into account.  For example, if one of the arguments is zero, I might not want to return a SumSpeciesElement, but rather the other summand.

However, this smells badly like duplicating a lot of code.

Reply:
Rethinking this: I guess it happens next to never that we are in one of the cases treated specially by `LazyModuleElement._add_ ` (or `_mul_`, etc.) So maybe my afterthought is thinking about the empty set.

Reply:
Does the following look correct to you?  At least it works.  I admit that the arguments `super` requires are not super-clear to me.
```
class LazySpeciesElement(LazyCompletionGradedAlgebraElement):
...
    def _add_(self, other):
        return SumSpeciesElement(self, other)
...
class SumSpeciesElement(LazySpeciesElement):
    def __init__(self, left, right):
        self._left = left
        self._right = right
        coeff_stream = super(LazySpeciesElement, type(left))._add_(left, right)._coeff_stream
        super().__init__(self.parent(), coeff_stream)

    def structures(self, labels):
        yield from self._left.structures(labels)
        yield from self._right.structures(labels)
```